Micronaut data / JPA - Multiple Datasource : Crash at start up

### Expected Behavior

I want to connect to two different data sources with mirconaut data and JPA (same database but different schemas). Bellow are the configurations (not relevant but it is an example).

```java
datasources:
  cat-database:
    url: jdbc:h2:mem:poc-cat;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
    driverClassName: org.h2.Driver
    username: sa
    password: ""
    dialect: H2
    auto-commit: false
  dog-database:
    url: jdbc:h2:mem:poc-dog;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
    driverClassName: org.h2.Driver
    username: sa
    password: ""
    dialect: H2
    auto-commit: false

jpa:
  cat-database:
    properties:
      hibernate.hbm2ddl.auto: update
      hibernate.dialect: org.hibernate.dialect.H2Dialect
      hibernate.show_sql: true
      hibernate.format_sql: true
      hibernate.default_schema: poc-cat
  dog-database:
    properties:
      hibernate.hbm2ddl.auto: update
      hibernate.dialect: org.hibernate.dialect.H2Dialect
      hibernate.show_sql: true
      hibernate.format_sql: true
      hibernate.default_schema: poc-dog
``` 

I have setup the following repositories :
For the "cat" database I have this one 
```java
@Repository("cat-database")
@Transactional("cat-database")
public interface CatRepository extends CrudRepository<Pet, Long> {
}
``` 

And for the "dog" database i have this one :
```java
@Repository("dog-database")
@Transactional("dog-database")
public interface DogRepository extends CrudRepository<Pet, Long> {
}
``` 

According to [the documentation](https://micronaut-projects.github.io/micronaut-data/latest/guide/index.html) my application should start smoothly.

### Actual Behaviour

There is the following crash at startup

```
 __  __ _                                  _   
|  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_ 
| |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
| |  | | | (__| | | (_) | | | | (_| | |_| | |_ 
|_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
16:44:55.425 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Bean definition [org.hibernate.SessionFactory] could not be loaded: Failed to inject value for parameter [jpaConfiguration] of class: io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory

Message: Multiple possible bean candidates found: [JpaConfiguration, JpaConfiguration]
Path Taken: 
@i.m.c.a.Context o.h.SessionFactory i.m.c.h.j.c.SessionFactoryPerDataSourceFactory.buildHibernateSessionFactoryBuilder#buildHibernateSessionFactoryBuilder(SessionFactoryBuilder sessionFactoryBuilder)
\---> new i.m.c.h.j.c.SessionFactoryPerDataSourceFactory(Environment environment, List<SessionFactoryBuilderConfigurer> configures, StandardServiceRegistryBuilderCreator serviceRegistryBuilderSupplier, List<StandardServiceRegistryBuilderConfigurer> standardServiceRegistryBuilderConfigurers, [@Primary JpaConfiguration jpaConfiguration], ApplicationContext applicationContext, @Primary Integrator integrator)
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [org.hibernate.SessionFactory] could not be loaded: Failed to inject value for parameter [jpaConfiguration] of class: io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory

Message: Multiple possible bean candidates found: [JpaConfiguration, JpaConfiguration]
Path Taken: 
@i.m.c.a.Context o.h.SessionFactory i.m.c.h.j.c.SessionFactoryPerDataSourceFactory.buildHibernateSessionFactoryBuilder#buildHibernateSessionFactoryBuilder(SessionFactoryBuilder sessionFactoryBuilder)
\---> new i.m.c.h.j.c.SessionFactoryPerDataSourceFactory(Environment environment, List<SessionFactoryBuilderConfigurer> configures, StandardServiceRegistryBuilderCreator serviceRegistryBuilderSupplier, List<StandardServiceRegistryBuilderConfigurer> standardServiceRegistryBuilderConfigurers, [@Primary JpaConfiguration jpaConfiguration], ApplicationContext applicationContext, @Primary Integrator integrator)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:2040)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:323)
	at io.micronaut.context.DefaultBeanContext.configureAndStartContext(DefaultBeanContext.java:3350)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:353)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:225)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:347)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:333)
	at com.example.poc.Application.main(Application.java:8)
Caused by: io.micronaut.context.exceptions.DependencyInjectionException: Failed to inject value for parameter [jpaConfiguration] of class: io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory

Message: Multiple possible bean candidates found: [JpaConfiguration, JpaConfiguration]
Path Taken: 
@i.m.c.a.Context o.h.SessionFactory i.m.c.h.j.c.SessionFactoryPerDataSourceFactory.buildHibernateSessionFactoryBuilder#buildHibernateSessionFactoryBuilder(SessionFactoryBuilder sessionFactoryBuilder)
\---> new i.m.c.h.j.c.SessionFactoryPerDataSourceFactory(Environment environment, List<SessionFactoryBuilderConfigurer> configures, StandardServiceRegistryBuilderCreator serviceRegistryBuilderSupplier, List<StandardServiceRegistryBuilderConfigurer> standardServiceRegistryBuilderConfigurers, [@Primary JpaConfiguration jpaConfiguration], ApplicationContext applicationContext, @Primary Integrator integrator)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2147)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1360)
	at io.micronaut.configuration.hibernate.jpa.conf.$SessionFactoryPerDataSourceFactory$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2337)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3150)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:3000)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2762)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1781)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1763)
	at io.micronaut.configuration.hibernate.jpa.conf.$SessionFactoryPerDataSourceFactory$BuildHibernateSessionFactoryBuilder4$Definition.instantiate(Unknown Source)
	at io.micronaut.context.BeanDefinitionDelegate.instantiate(BeanDefinitionDelegate.java:183)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2337)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3150)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.intializeEagerBean(DefaultBeanContext.java:3039)
	at io.micronaut.context.DefaultBeanContext.initializeEagerBean(DefaultBeanContext.java:2699)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:2034)
	... 8 common frames omitted
Caused by: io.micronaut.context.exceptions.NonUniqueBeanException: Multiple possible bean candidates found: [JpaConfiguration, JpaConfiguration]
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidate(DefaultBeanContext.java:2458)
	at io.micronaut.context.DefaultApplicationContext.findConcreteCandidate(DefaultApplicationContext.java:675)
	at io.micronaut.context.DefaultBeanContext.lastChanceResolve(DefaultBeanContext.java:3323)
	at io.micronaut.context.DefaultBeanContext.pickOneBean(DefaultBeanContext.java:3257)
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidateNoCache(DefaultBeanContext.java:3213)
	at io.micronaut.context.DefaultBeanContext.findConcreteCandidate(DefaultBeanContext.java:3196)
	at io.micronaut.context.DefaultBeanContext.findBeanDefinition(DefaultBeanContext.java:2783)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2752)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1844)
	at io.micronaut.context.AbstractBeanResolutionContext.findBean(AbstractBeanResolutionContext.java:286)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2127)
	... 25 common frames omitted

Process finished with exit code 1
``` 

### Steps To Reproduce

1. Setup an application with Micronaut Data JPA as dependency 
2. Use the configuration I provided earlier
3. Start the application

### Environment Information

- Operating system : linux ubuntu
- JDK : 21 (temurin)
- Micronaut platform : 4.9.2

### Example Application

_No response_

### Version

4.9.2

--- a/hibernate-jpa/src/main/java/io/micronaut/configuration/hibernate/jpa/conf/SessionFactoryPerDataSourceFactory.java
+++ b/hibernate-jpa/src/main/java/io/micronaut/configuration/hibernate/jpa/conf/SessionFactoryPerDataSourceFactory.java
@@ -41,6 +41,8 @@ import org.hibernate.service.ServiceRegistry;
 
 import javax.sql.DataSource;
 import java.util.List;
+import jakarta.inject.Inject;
+import io.micronaut.inject.qualifiers.Qualifiers;
 
 /**
  * Factory that builds {@link SessionFactory} per {@link DataSource}.
@@ -56,15 +58,40 @@ final class SessionFactoryPerDataSourceFactory extends AbstractHibernateFactory 
 
     private final JpaConfiguration defaultJpaConfiguration;
 
+    /**
+     * Backward-compatible constructor that previously injected a primary JpaConfiguration.
+     * Deprecated: prefer the new constructor without JpaConfiguration to avoid DI ambiguity when multiple JPA configs exist.
+     */
+    @Deprecated
     SessionFactoryPerDataSourceFactory(Environment environment,
                                        List<SessionFactoryBuilderConfigurer> configures,
                                        StandardServiceRegistryBuilderCreator serviceRegistryBuilderSupplier,
                                        List<StandardServiceRegistryBuilderConfigurer> standardServiceRegistryBuilderConfigurers,
-                                       @Primary @Nullable JpaConfiguration jpaConfiguration,
+                                       @Primary @Nullable JpaConfiguration ignoredJpaConfiguration,
+                                       ApplicationContext applicationContext,
+                                       @Primary @Nullable Integrator integrator) {
+        this(environment, configures, serviceRegistryBuilderSupplier, standardServiceRegistryBuilderConfigurers, applicationContext, integrator);
+    }
+
+    /**
+     * Preferred constructor: constructs a neutral fallback JpaConfiguration and avoids injecting a primary bean.
+     */
+    @Inject
+    SessionFactoryPerDataSourceFactory(Environment environment,
+                                       List<SessionFactoryBuilderConfigurer> configures,
+                                       StandardServiceRegistryBuilderCreator serviceRegistryBuilderSupplier,
+                                       List<StandardServiceRegistryBuilderConfigurer> standardServiceRegistryBuilderConfigurers,
                                        ApplicationContext applicationContext,
                                        @Primary @Nullable Integrator integrator) {
         super(environment, configures, serviceRegistryBuilderSupplier, standardServiceRegistryBuilderConfigurers);
-        this.defaultJpaConfiguration = jpaConfiguration != null ? jpaConfiguration : new JpaConfiguration(applicationContext, integrator);
+        JpaConfiguration def = applicationContext.findBean(JpaConfiguration.class, Qualifiers.byName(JpaConfiguration.PRIMARY)).orElse(null);
+        if (def == null) {
+            var all = applicationContext.getBeansOfType(JpaConfiguration.class);
+            if (all.size() == 1) {
+                def = all.iterator().next();
+            }
+        }
+        this.defaultJpaConfiguration = def != null ? def : new JpaConfiguration(applicationContext, integrator);
     }
 
     @EachBean(DataSource.class)

--- a/hibernate-jpa/src/main/java/io/micronaut/configuration/hibernate/jpa/conf/SessionFactoryPerDataSourceFactory.java
+++ b/hibernate-jpa/src/main/java/io/micronaut/configuration/hibernate/jpa/conf/SessionFactoryPerDataSourceFactory.java
@@ -41,7 +41,6 @@ import org.hibernate.service.ServiceRegistry;
 
 import javax.sql.DataSource;
 import java.util.List;
-import jakarta.inject.Inject;
 import io.micronaut.inject.qualifiers.Qualifiers;
 
 /**
@@ -58,25 +57,6 @@ final class SessionFactoryPerDataSourceFactory extends AbstractHibernateFactory 
 
     private final JpaConfiguration defaultJpaConfiguration;
 
-    /**
-     * Backward-compatible constructor that previously injected a primary JpaConfiguration.
-     * Deprecated: prefer the new constructor without JpaConfiguration to avoid DI ambiguity when multiple JPA configs exist.
-     */
-    @Deprecated
-    SessionFactoryPerDataSourceFactory(Environment environment,
-                                       List<SessionFactoryBuilderConfigurer> configures,
-                                       StandardServiceRegistryBuilderCreator serviceRegistryBuilderSupplier,
-                                       List<StandardServiceRegistryBuilderConfigurer> standardServiceRegistryBuilderConfigurers,
-                                       @Primary @Nullable JpaConfiguration ignoredJpaConfiguration,
-                                       ApplicationContext applicationContext,
-                                       @Primary @Nullable Integrator integrator) {
-        this(environment, configures, serviceRegistryBuilderSupplier, standardServiceRegistryBuilderConfigurers, applicationContext, integrator);
-    }
-
-    /**
-     * Preferred constructor: constructs a neutral fallback JpaConfiguration and avoids injecting a primary bean.
-     */
-    @Inject
     SessionFactoryPerDataSourceFactory(Environment environment,
                                        List<SessionFactoryBuilderConfigurer> configures,
                                        StandardServiceRegistryBuilderCreator serviceRegistryBuilderSupplier,

--- a/hibernate-jpa/src/test/groovy/io/micronaut/configuration/hibernate/jpa/Cat.java
+++ b/hibernate-jpa/src/test/groovy/io/micronaut/configuration/hibernate/jpa/Cat.java
@@ -1,0 +1,19 @@
+package io.micronaut.configuration.hibernate.jpa;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+
+@Entity
+public class Cat {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+    private String name;
+
+    public Long getId() { return id; }
+    public void setId(Long id) { this.id = id; }
+    public String getName() { return name; }
+    public void setName(String name) { this.name = name; }
+}

--- a/hibernate-jpa/src/test/groovy/io/micronaut/configuration/hibernate/jpa/MultiJpaPerDatasourceSpec.groovy
+++ b/hibernate-jpa/src/test/groovy/io/micronaut/configuration/hibernate/jpa/MultiJpaPerDatasourceSpec.groovy
@@ -1,0 +1,73 @@
+package io.micronaut.configuration.hibernate.jpa
+
+import io.micronaut.context.ApplicationContext
+import org.hibernate.SessionFactory
+import spock.lang.Specification
+
+class MultiJpaPerDatasourceSpec extends Specification {
+
+    void "two datasources with two jpa configs start and create two session factories"() {
+        given:
+        def ctx = ApplicationContext.run([
+                'datasources.cat-database.url'             : 'jdbc:h2:mem:poc-cat;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE',
+                'datasources.cat-database.driverClassName' : 'org.h2.Driver',
+                'datasources.cat-database.username'        : 'sa',
+                'datasources.cat-database.password'        : '',
+                'datasources.cat-database.dialect'         : 'H2',
+
+                'datasources.dog-database.url'             : 'jdbc:h2:mem:poc-dog;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE',
+                'datasources.dog-database.driverClassName' : 'org.h2.Driver',
+                'datasources.dog-database.username'        : 'sa',
+                'datasources.dog-database.password'        : '',
+                'datasources.dog-database.dialect'         : 'H2',
+
+                'jpa.cat-database.properties.hibernate.hbm2ddl.auto': 'update',
+                'jpa.cat-database.properties.hibernate.dialect'     : 'org.hibernate.dialect.H2Dialect',
+                'jpa.cat-database.packages-to-scan'                 : 'io.micronaut.configuration.hibernate.jpa',
+
+                'jpa.dog-database.properties.hibernate.hbm2ddl.auto': 'update',
+                'jpa.dog-database.properties.hibernate.dialect'     : 'org.hibernate.dialect.H2Dialect',
+                'jpa.dog-database.packages-to-scan'                 : 'io.micronaut.configuration.hibernate.jpa',
+        ])
+
+        when:
+        def beans = ctx.getBeansOfType(SessionFactory)
+
+        then:
+        beans.size() == 2
+
+        cleanup:
+        ctx?.close()
+    }
+
+    void "two datasources with one jpa config fallback copy produces two session factories"() {
+        given:
+        def ctx = ApplicationContext.run([
+                'datasources.cat-database.url'             : 'jdbc:h2:mem:poc-cat;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE',
+                'datasources.cat-database.driverClassName' : 'org.h2.Driver',
+                'datasources.cat-database.username'        : 'sa',
+                'datasources.cat-database.password'        : '',
+                'datasources.cat-database.dialect'         : 'H2',
+
+                'datasources.dog-database.url'             : 'jdbc:h2:mem:poc-dog;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE',
+                'datasources.dog-database.driverClassName' : 'org.h2.Driver',
+                'datasources.dog-database.username'        : 'sa',
+                'datasources.dog-database.password'        : '',
+                'datasources.dog-database.dialect'         : 'H2',
+
+                // Only one jpa config; entity scan covers both
+                'jpa.cat-database.properties.hibernate.hbm2ddl.auto': 'update',
+                'jpa.cat-database.properties.hibernate.dialect'     : 'org.hibernate.dialect.H2Dialect',
+                'jpa.cat-database.packages-to-scan'                 : 'io.micronaut.configuration.hibernate.jpa'
+        ])
+
+        when:
+        def beans = ctx.getBeansOfType(SessionFactory)
+
+        then:
+        beans.size() == 2
+
+        cleanup:
+        ctx?.close()
+    }
+}

